### ISSUE TITLE
[v0.8.25] CosyVoice-TTS - long delay, pitch shifting and accent issues

### ISSUE THREAD

**Issue by subof** [CrispStrobe/CrispASR#334]
Hi,
I am experiencing issues with voice cloning using the cosyvoice3-tts backend (model: cosyvoice3-0.5b-2512.gguf) integrated into Subtitle Edit. The output pitch and cloning quality change drastically depending on the reference audio sample rate, suggesting an issue with sample rate handling or resampling in the CrispASR backend.
Observed behavior:

   1. With 16.0 kHz Reference Audio:
   * Specs: PCM, 16.0 kHz, 16 bits, 1 channel, 256 kb/s, ~7s duration.
      * Result: The cloned voice has a very heavy, unnatural foreign accent, likely because the 16 kHz audio is fed into the 24 kHz model without proper resampling, corrupting phoneme alignment.
   2. With 24.0 kHz Reference Audio:
   * Specs: PCM, 24.0 kHz, 16 bits, 1 channel, 384 kb/s, ~20s duration.
      * Result: Pronunciation improves, but a severe "chipmunk effect" occurs. The output pitch is significantly shifted upward, which points to a mismatch in the output vocoder sample rate or time-step scaling inside the C++ container.
   
Expected behavior:
The backend should automatically resample input reference audio to the model's native sample rate and generate the output voice with the correct pitch and speed, regardless of the input file's parameters.
Could you please check how the sample rate and time-steps are managed for the CosyVoice3 backend? Thank you!

I tested it on Windows, HP Z8 G4 RTX 3090TI

**Comment by subof** [comment]
p.s. The basic model Fun-CosyVoice3-0.5B-2512 is the standard pre-trained version, while the RL modification has undergone additional Reinforcement Learning (RL).
The RL version is designed specifically to achieve maximum speech quality, accuracy, and generation stability.

Could you also add the RL version to the backend for better voicing, please?
Thank you.

**Comment by CrispStrobe** [comment]
Thanks — this reproduced, though the cause turned out not to be the sample rate. Four things are fixed on `main`, and the RL model is published.

**The sample rate is handled correctly.** I checked the reference front-end against upstream's own `speech_tokenizer_v3.onnx`: our C++ speech tokens are byte-exact (201/201) whether the reference arrives natively at 16 kHz or is resampled there from 24/32/48 kHz, and our resampler tracks torchaudio. A single recording fed in at 8, 16, 22.05, 24, 32, 44.1 and 48 kHz now produces the same transcript at the same speaking rate. (Output is always 24 kHz mono.)

**What actually breaks it is `--ref-text` not matching the reference audio.** The talker is conditioned on the pair *(reference transcript, reference speech)* and infers the speaker's rate from it. Give it a 20 s clip labelled with a partial or approximate transcript and it concludes the speaker takes 20 s to say one sentence — then it either stops immediately (`AR decode produced 0 tokens`, no audio) or crams your line into far too few 40 ms frames, which is exactly the "chipmunk effect". Your 20 s reference is much harder to transcribe exactly than the 7 s one, which fits the difference you saw between the two files.

Fixes:

1. **Minimum decode length.** Upstream masks the stop token until `2 × (target text tokens)` speech tokens are out (`min_token_text_ratio` in `Qwen2LM.inference`, which `CosyVoice3LM` inherits); we had no such floor, so one unlucky sample at step 0 ended the decode. Ported. It also bounds how fast the model may speak, which is the direct anti-chipmunk guard.
2. **A warning when the transcript can't match the clip.** The runtime now checks the pairing against the same 2–20 speech-tokens-per-text-token band the decode works in, and tells you what to do:
   ```
   cosyvoice3_tts: WARNING: the reference clip holds 17.72 s of speech but --ref-text is
   only 7 token(s) long (63.3 speech frames per text token; a matching transcript lands
   between 2 and 20). ...
   ```
3. **The "long delay".** The speaker was re-extracted from the WAV on *every* synthesis — speech tokenizer + CAMPPlus + prompt mel — and the CLI synthesises once per sentence, plus once more for the spoken AI disclaimer. Now cached per reference: a three-sentence request went 66.0 s → 45.7 s with byte-identical output.
4. A real (if small) defect in the polyphase resampler: it truncated its filter on every downsample. Fixed, with a per-sample regression test.

**Practical advice for your clips:** `--ref-text` must be a complete, exact transcription of the reference audio. If you can't transcribe all 20 s, **trim the clip to the part you did transcribe** — a clean 4–10 s excerpt clones better than 20 s with an approximate transcript.

**RL model: added.** `llm.rl.pt` is now converted and published alongside the base talker in [`cstr/cosyvoice3-0.5b-2512-GGUF`](https://huggingface.co/cstr/cosyvoice3-0.5b-2512-GGUF) as `cosyvoice3-llm-rl-f16.gguf` / `cosyvoice3-llm-rl-q4_k.gguf`. Only the talker differs — flow, HiFT, CAMPPlus, the speech tokenizer and the voice bank are shared — so it's one 384 MB swap:

```
crispasr --backend cosyvoice3-tts-rl -m auto \
    --voice ref.wav --ref-text "<exact transcript>" --i-have-rights \
    --tts "..." --tts-output out.wav
```

`cosyvoice3-tts-rl` (aliases `cosyvoice3-rl`, `cv3-rl`) is the same engine wired so `-m auto` fetches the RL talker and reuses the shared companions; pointing `-m` at the GGUF directly works too. Both talkers round-trip correctly through ASR here, on bank voices and on WAV cloning.

This will be in the next release; until then it needs a build from `main`. Details in [`docs/tts.md`](https://github.com/CrispStrobe/CrispASR/blob/main/docs/tts.md) under "CosyVoice3 — voice cloning from a WAV".

**Comment by CrispStrobe** [comment]
Follow-up: **`--ref-text` is now optional for `cosyvoice3-tts`**, which removes the trap I described above rather than just warning about it.

If you leave it off, the reference clip is transcribed for you (whisper by default, `--ref-asr <backend>` to choose another) and the result is cached beside the clip as `<voice>.cv3reftext`, so only the first run pays for it. An explicit `--ref-text` still wins and skips ASR entirely.

```
crispasr --backend cosyvoice3-tts -m auto \
    --voice ref.wav --i-have-rights \
    --tts "Text to speak." --tts-output out.wav
```

Measured on a 17.7 s reference, asking for "The quick brown fox jumps over the lazy dog near the river bank.":

| reference transcript | result |
|---|---|
| a plausible one-sentence guess | warning fires, and the requested line is **missing** from the output entirely |
| omitted → auto-transcribed | the requested line comes out in full |

So for your ~20 s clip: just drop `--ref-text`. Trimming to a clean 4–10 s excerpt is still a good idea for clone quality, but it is no longer needed to avoid this failure.

One limitation worth stating: this is the CLI and server path. If you drive CrispASR through the C/Python/Rust session API, `set_voice()` still requires a transcript for a WAV — that layer can't spin up an ASR model for itself yet.

On `main` as of 7b6bcfe9.

**Comment by CrispStrobe** [comment]
One more, and this one is the likeliest cause of the *"very heavy, unnatural foreign accent"* you described.

While adding diff-harness coverage for the voice-clone front end I found that **the speaker embedding was being computed wrong** — cos **0.737** against upstream's `campplus.onnx`, on every `--voice ref.wav` clone.

CAMPPlus ends with

```
transit3.linear (Conv1d, bias=False) → out_nonlinear (BatchNorm + ReLU) → StatsPool
```

and CosyVoice3's ONNX export **folds that trailing BatchNorm back into the convolution**: the graph has a bare `/xvector/out_nonlinear/relu/Relu` with no BN parameters anywhere, while `/xvector/transit3/linear/Conv` gained a fused weight *and* a fused bias. Our code never applied a bias on that convolution, so the fold was silently dropped and the last 512 channels reached the statistics pooling un-normalised. (Zeroing that same bias in the ONNX reference reproduces our old output at cos 0.999998, which is what pins it to this one term.) Fixed: **0.999997**.

This is why it read as "cloning quality" rather than a bug — a wrong-but-plausible 192-dimensional vector just clones a slightly wrong voice. It also explains why the **baked bank voices** (`--voice fleurs-en` etc.) always sounded fine: their embeddings are computed by the ONNX model in Python at build time and shipped in `cosyvoice3-voices.gguf`. Only the runtime WAV path went through the broken code.

Measured effect on a clone, speaker similarity to the reference (Resemblyzer): 0.7245 → 0.7537.

The three clone-front-end stages now have permanent coverage in `crispasr-diff`:

```
[PASS] clone_spk_emb          shape=[192]      cos=0.999996
[PASS] clone_prompt_feat_24k  shape=[80,213]   cos_mean=0.999948
[PASS] clone_speech_tokens    shape=[107]      cos=1.000000
```

On `main` as of a629ad06. Worth re-testing your clips once you can build from `main` — between this, the decode floor and the auto-transcribed reference, all three of the symptoms you reported have had a distinct cause fixed.

**Comment by CrispStrobe** [comment]
Closing the loop on the one limitation left open above.

> If you drive CrispASR through the C/Python/Rust session API, `set_voice()` still requires a transcript for a WAV — that layer can't spin up an ASR model for itself yet.

Still true that it cannot transcribe — the CLI helper constructs a second `CrispasrBackend`, which the session ABI has no access to. But it does not need to: the CLI already caches its auto-transcript beside the clip, and that cache lives in `core/` precisely so every consumer can share it. So as of 97412387, a clip prepared once through the CLI clones through the session API too, instead of failing on a transcript that has already been computed and is sitting right next to the file.

Three arms, measured with `cosyvoice3-llm-q4_k`:

| | |
|---|---|
| session `set_voice(wav)`, no cache | `-2`, now with an explanation |
| CLI run without `--ref-text` | auto-transcribes, caches 79 bytes |
| same session call, unchanged | succeeds — 2.36 s of audio, ASR round-trip returns the requested line exactly |

When there is genuinely no transcript, the bare `-2` is now an explanation rather than a number. That matters more here than usual, because as you experienced, an *approximate* transcript is worse than none — the talker infers the speaker's rate from it and rushes or truncates the line. So the message says to pass the exact transcript or let the CLI produce one, rather than implying any string will do.

One note on how this was built, since it is a small lesson: the suffix is now a single shared constant used by both layers rather than a string literal on each side. My first regression test declared its own copy of that literal, which made it self-consistent and therefore blind — changing the suffix moved writer and reader together and every assertion still passed. Red-verifying caught it, and a shared constant removes the drift rather than testing for it.

@subof — your original three symptoms all have distinct causes fixed on `main` (the CAMPPlus bias fold for the accent, the decode floor for the chipmunk effect, the per-reference cache for the delay), and the RL talker is published. **Closing on the strength of the measurements in this thread rather than a retest**, since it has been a while — please reopen if your clips still misbehave on a current build, and if you do, the single most useful thing is to drop `--ref-text` entirely and let it transcribe the reference for you.
